File: src/jiraClient.js
/**
 * Jira API v3 Client
 * Handles all interactions with Jira REST API v3
 */

export class JiraClient {
  constructor(config) {
    this.host = config.host;
    this.email = config.email;
    this.apiToken = config.apiToken;
    this.baseUrl = `https://${this.host}/rest/api/3`;
    this.agileUrl = `https://${this.host}/rest/agile/1.0`;
  }

  /**
   * Make authenticated request to Jira API
   */
  async request(url, options = {}) {
    const auth = Buffer.from(`${this.email}:${this.apiToken}`).toString('base64');

    const response = await fetch(url, {
      ...options,
      headers: {
        'Authorization': `Basic ${auth}`,
        'Accept': 'application/json',
        'Content-Type': 'application/json',
        ...options.headers,
      },
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Jira API Error (${response.status}): ${errorText}`);
    }

    return response.json();
  }

  /**
   * Get all boards for a project
   */
  async getBoards(projectKey) {
    const url = `${this.agileUrl}/board?projectKeyOrId=${projectKey}`;
    const data = await this.request(url);
    return data.values;
  }

  /**
   * Get all boards for multiple projects
   */
  async getAllBoardsForProjects(projectKeys) {
    const allBoards = [];
    for (const projectKey of projectKeys) {
      try {
        const boards = await this.getBoards(projectKey);
        allBoards.push(...boards.map(board => ({ ...board, projectKey })));
      } catch (error) {
        console.warn(`Warning: Could not fetch boards for project ${projectKey}: ${error.message}`);
      }
    }
    return allBoards;
  }

  /**
   * Get all sprints for a board
   */
  async getSprints(boardId) {
    const url = `${this.agileUrl}/board/${boardId}/sprint`;
    const data = await this.request(url);
    return data.values;
  }

  /**
   * Get the most recently closed sprint for a team across all projects
   * Searches all boards in all projects for sprints with the team name
   */
  async getLastClosedSprintForTeam(projectKeys, teamLabel) {
    const allBoards = await this.getAllBoardsForProjects(projectKeys);

    if (allBoards.length === 0) {
      throw new Error(`No boards found for projects: ${projectKeys.join(', ')}`);
    }

    let allMatchingSprints = [];

    // Search all boards for sprints with the team name
    for (const board of allBoards) {
      try {
        const sprints = await this.getSprints(board.id);

        // Filter for closed sprints with team name in sprint name
        const teamSprints = sprints.filter(sprint =>
          sprint.state === 'closed' &&
          sprint.name.toLowerCase().includes(teamLabel.toLowerCase())
        );

        // Add board and project info to each sprint
        teamSprints.forEach(sprint => {
          allMatchingSprints.push({
            sprint,
            boardId: board.id,
            boardName: board.name,
            projectKey: board.projectKey
          });
        });
      } catch (error) {
        console.warn(`Warning: Could not fetch sprints for board ${board.name}: ${error.message}`);
      }
    }

    if (allMatchingSprints.length === 0) {
      throw new Error(`No closed sprints found with team name "${teamLabel}" across projects: ${projectKeys.join(', ')}`);
    }

    // Sort by end date to get most recent
    allMatchingSprints.sort((a, b) =>
      new Date(b.sprint.endDate) - new Date(a.sprint.endDate)
    );

    const mostRecent = allMatchingSprints[0];
    return {
      sprint: mostRecent.sprint,
      boardId: mostRecent.boardId,
      boardName: mostRecent.boardName,
      projectKey: mostRecent.projectKey
    };
  }

  /**
   * Get the most recently closed sprint for a project, optionally filtered by team name
   */
  async getLastClosedSprint(projectKey, teamLabel = null) {
    const boards = await this.getBoards(projectKey);
    if (boards.length === 0) {
      throw new Error(`No boards found for project ${projectKey}`);
    }

    // Use the first board (or you can add logic to select specific board)
    const boardId = boards[0].id;
    const sprints = await this.getSprints(boardId);

    // Filter for closed sprints
    let closedSprints = sprints.filter(sprint => sprint.state === 'closed');

    // If team label provided, try to find sprints matching the team name
    if (teamLabel) {
      const teamSpecificSprints = closedSprints.filter(sprint =>
        sprint.name.toLowerCase().includes(teamLabel.toLowerCase())
      );

      // If we found team-specific sprints, use those; otherwise fall back to all sprints
      if (teamSpecificSprints.length > 0) {
        closedSprints = teamSpecificSprints;
      }
    }

    // Sort by end date to get most recent
    closedSprints.sort((a, b) => new Date(b.endDate) - new Date(a.endDate));

    if (closedSprints.length === 0) {
      throw new Error(`No closed sprints found for project ${projectKey}`);
    }

    return { sprint: closedSprints[0], boardId, projectKey };
  }

  /**
   * Get the most recently closed sprints for multiple projects
   */
  async getLastClosedSprintsForProjects(projectKeys) {
    const sprints = [];
    for (const projectKey of projectKeys) {
      try {
        const sprintData = await this.getLastClosedSprint(projectKey);
        sprints.push(sprintData);
      } catch (error) {
        console.warn(`Warning: ${error.message}`);
      }
    }
    return sprints;
  }

  /**
   * Get all issues in a sprint
   */
  async getSprintIssues(sprintId, teamLabel = null) {
    const url = `${this.agileUrl}/sprint/${sprintId}/issue?maxResults=1000`;
    const data = await this.request(url);
    let issues = data.issues;

    // Filter by team label if specified
    if (teamLabel) {
      issues = issues.filter(issue =>
        issue.fields.labels && issue.fields.labels.includes(teamLabel)
      );
    }

    return issues;
  }

  /**
   * Get unique team labels from sprint issues
   */
  async getTeamLabelsFromSprint(sprintId) {
    const url = `${this.agileUrl}/sprint/${sprintId}/issue?maxResults=1000`;
    const data = await this.request(url);
    const labelsSet = new Set();

    data.issues.forEach(issue => {
      if (issue.fields.labels) {
        issue.fields.labels.forEach(label => labelsSet.add(label));
      }
    });

    return Array.from(labelsSet);
  }

  /**
   * Get detailed issue information including changelog
   */
  async getIssueDetails(issueKey) {
    const url = `${this.baseUrl}/issue/${issueKey}?expand=changelog`;
    return this.request(url);
  }

  /**
   * Get users assigned to issues in the sprint
   */
  async getSprintTeamMembers(issues) {
    const uniqueUsers = new Map();

    for (const issue of issues) {
      if (issue.fields.assignee) {
        const user = issue.fields.assignee;
        if (!uniqueUsers.has(user.accountId)) {
          uniqueUsers.set(user.accountId, {
            accountId: user.accountId,
            displayName: user.displayName,
            emailAddress: user.emailAddress || 'N/A',
            avatarUrl: user.avatarUrls?.['48x48'] || '',
          });
        }
      }
    }

    return Array.from(uniqueUsers.values());
  }

  /**
   * Get project details
   */
  async getProject(projectKey) {
    const url = `${this.baseUrl}/project/${projectKey}`;
    return this.request(url);
  }

  /**
   * Search for issues with JQL
   */
  async searchIssues(jql, fields = ['summary', 'status', 'assignee']) {
    const url = `${this.baseUrl}/search`;
    const body = {
      jql,
      fields,
      maxResults: 1000,
    };

    return this.request(url, {
      method: 'POST',
      body: JSON.stringify(body),
    });
  }
}

File: sprint_summary_agent/jira_client.py
"""Jira API v3 Client for interacting with Jira REST API."""

import base64
from datetime import datetime
from typing import Any, Dict, List, Optional

import requests


class JiraClient:
    """Handles all interactions with Jira REST API v3."""

    def __init__(self, config: dict):
        """Initialize Jira client with configuration.

        Args:
            config: Dictionary with 'host', 'email', and 'api_token' keys
        """
        self.host = config["host"]
        self.email = config["email"]
        self.api_token = config["api_token"]
        self.base_url = f"https://{self.host}/rest/api/3"
        self.agile_url = f"https://{self.host}/rest/agile/1.0"

    def _get_auth_header(self) -> str:
        """Generate Basic Auth header."""
        credentials = f"{self.email}:{self.api_token}"
        encoded = base64.b64encode(credentials.encode()).decode()
        return f"Basic {encoded}"

    def request(self, url: str, method: str = "GET", **kwargs) -> Dict[str, Any]:
        """Make authenticated request to Jira API.

        Args:
            url: Full URL to request
            method: HTTP method
            **kwargs: Additional arguments for requests

        Returns:
            JSON response data

        Raises:
            Exception: If request fails
        """
        headers = {
            "Authorization": self._get_auth_header(),
            "Accept": "application/json",
            "Content-Type": "application/json",
        }
        headers.update(kwargs.pop("headers", {}))

        response = requests.request(method, url, headers=headers, **kwargs)

        if not response.ok:
            raise Exception(f"Jira API Error ({response.status_code}): {response.text}")

        return response.json()

    def get_boards(self, project_key: str) -> List[Dict[str, Any]]:
        """Get all boards for a project."""
        url = f"{self.agile_url}/board?projectKeyOrId={project_key}"
        data = self.request(url)
        return data.get("values", [])

    def get_all_boards_for_projects(self, project_keys: List[str]) -> List[Dict[str, Any]]:
        """Get all boards for multiple projects."""
        all_boards = []
        for project_key in project_keys:
            try:
                boards = self.get_boards(project_key)
                for board in boards:
                    board["projectKey"] = project_key
                    all_boards.append(board)
            except Exception as e:
                print(f"Warning: Could not fetch boards for project {project_key}: {e}")
        return all_boards

    def get_sprints(self, board_id: int) -> List[Dict[str, Any]]:
        """Get all sprints for a board."""
        url = f"{self.agile_url}/board/{board_id}/sprint"
        data = self.request(url)
        return data.get("values", [])

    def get_last_closed_sprint_for_team(
        self, project_keys: List[str], team_label: str
    ) -> Dict[str, Any]:
        """Get the most recently closed sprint for a team across all projects.

        Args:
            project_keys: List of project keys to search
            team_label: Team name to search for in sprint names

        Returns:
            Dictionary with sprint, boardId, boardName, and projectKey

        Raises:
            Exception: If no sprints found
        """
        all_boards = self.get_all_boards_for_projects(project_keys)

        if not all_boards:
            raise Exception(f"No boards found for projects: {', '.join(project_keys)}")

        all_matching_sprints = []

        # Search all boards for sprints with the team name
        for board in all_boards:
            try:
                sprints = self.get_sprints(board["id"])

                # Filter for closed sprints with team name in sprint name
                team_sprints = [
                    sprint
                    for sprint in sprints
                    if sprint.get("state") == "closed"
                    and team_label.lower() in sprint.get("name", "").lower()
                ]

                # Add board and project info to each sprint
                for sprint in team_sprints:
                    all_matching_sprints.append(
                        {
                            "sprint": sprint,
                            "boardId": board["id"],
                            "boardName": board["name"],
                            "projectKey": board["projectKey"],
                        }
                    )
            except Exception as e:
                print(f"Warning: Could not fetch sprints for board {board['name']}: {e}")

        if not all_matching_sprints:
            raise Exception(
                f'No closed sprints found with team name "{team_label}" '
                f"across projects: {', '.join(project_keys)}"
            )

        # Sort by end date to get most recent
        all_matching_sprints.sort(
            key=lambda x: datetime.fromisoformat(x["sprint"]["endDate"].replace("Z", "+00:00")),
            reverse=True,
        )

        most_recent = all_matching_sprints[0]
        return {
            "sprint": most_recent["sprint"],
            "boardId": most_recent["boardId"],
            "boardName": most_recent["boardName"],
            "projectKey": most_recent["projectKey"],
        }

    def get_last_closed_sprint(
        self, project_key: str, team_label: Optional[str] = None
    ) -> Dict[str, Any]:
        """Get the most recently closed sprint for a project.

        Args:
            project_key: Project key
            team_label: Optional team label to filter by

        Returns:
            Dictionary with sprint, boardId, and projectKey

        Raises:
            Exception: If no sprints found
        """
        boards = self.get_boards(project_key)
        if not boards:
            raise Exception(f"No boards found for project {project_key}")

        # Use the first board
        board_id = boards[0]["id"]
        sprints = self.get_sprints(board_id)

        # Filter for closed sprints
        closed_sprints = [sprint for sprint in sprints if sprint.get("state") == "closed"]

        # If team label provided, try to find sprints matching the team name
        if team_label:
            team_specific_sprints = [
                sprint
                for sprint in closed_sprints
                if team_label.lower() in sprint.get("name", "").lower()
            ]

            # If we found team-specific sprints, use those
            if team_specific_sprints:
                closed_sprints = team_specific_sprints

        # Sort by end date to get most recent
        closed_sprints.sort(
            key=lambda x: datetime.fromisoformat(x["endDate"].replace("Z", "+00:00")),
            reverse=True,
        )

        if not closed_sprints:
            raise Exception(f"No closed sprints found for project {project_key}")

        return {"sprint": closed_sprints[0], "boardId": board_id, "projectKey": project_key}

    def get_sprint_issues(self, sprint_id: int, team_label: Optional[str] = None) -> List[Dict[str, Any]]:
        """Get all issues in a sprint.

        Args:
            sprint_id: Sprint ID
            team_label: Optional team label to filter by

        Returns:
            List of issues
        """
        url = f"{self.agile_url}/sprint/{sprint_id}/issue?maxResults=1000"
        data = self.request(url)
        issues = data.get("issues", [])

        # Filter by team label if specified
        if team_label:
            issues = [
                issue
                for issue in issues
                if issue.get("fields", {}).get("labels")
                and team_label in issue["fields"]["labels"]
            ]

        return issues

    def get_team_labels_from_sprint(self, sprint_id: int) -> List[str]:
        """Get unique team labels from sprint issues."""
        url = f"{self.agile_url}/sprint/{sprint_id}/issue?maxResults=1000"
        data = self.request(url)
        labels_set = set()

        for issue in data.get("issues", []):
            issue_labels = issue.get("fields", {}).get("labels", [])
            labels_set.update(issue_labels)

        return sorted(labels_set)

    def get_sprint_team_members(self, issues: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Get users assigned to issues in the sprint."""
        unique_users = {}

        for issue in issues:
            assignee = issue.get("fields", {}).get("assignee")
            if assignee:
                account_id = assignee.get("accountId")
                if account_id and account_id not in unique_users:
                    unique_users[account_id] = {
                        "accountId": account_id,
                        "displayName": assignee.get("displayName", "Unknown"),
                        "emailAddress": assignee.get("emailAddress", "N/A"),
                        "avatarUrl": assignee.get("avatarUrls", {}).get("48x48", ""),
                    }

        return list(unique_users.values())

    def get_project(self, project_key: str) -> Dict[str, Any]:
        """Get project details."""
        url = f"{self.base_url}/project/{project_key}"
        return self.request(url)

    def search_issues(
        self, jql: str, fields: Optional[List[str]] = None
    ) -> Dict[str, Any]:
        """Search for issues with JQL."""
        if fields is None:
            fields = ["summary", "status", "assignee"]

        url = f"{self.base_url}/search"
        body = {"jql": jql, "fields": fields, "maxResults": 1000}

        return self.request(url, method="POST", json=body)
